pkg/openstack/common.go

```diff
@@ -27,3 +44,301 @@ func EnsureDeleted(ctx context.Context, helper *helper.Helper, obj client.Object
+) *corev1.MetalLBConfig {
+	for _, metallbcfg := range externalEndpoints {
+		if metallbcfg.Endpoint == endpt {
+			return metallbcfg.DeepCopy()
+		}
+	}
+
+	return nil
+}
+
+// ServiceDetails - service details to create service.Service with overrides
+type ServiceDetails struct {
+	ServiceName         string
+	Namespace           string
+	Endpoint            service.Endpoint
+	ExternalEndpoints   []corev1.MetalLBConfig
+	ServiceOverrideSpec []service.OverrideSpec
+	RouteOverrideSpec   *route.OverrideSpec
+	endpointName        string
+	endpointURL         string
+	baseServiceLabels   map[string]string
+	serviceLabels       map[string]string
+	hostname            *string
+	route               *routev1.Route
+}
+
+// GetServiceLabels - returns the labels for the service, used as selector on the route
+func (sd *ServiceDetails) GetServiceLabels() map[string]string {
+	return sd.serviceLabels
+}
+
+// GetEndpointName - returns the name of the endpoint
+func (sd *ServiceDetails) GetEndpointName() string {
+	return sd.endpointName
+}
+
+// GetEndpointURL - returns the URL of the endpoint (proto://hostname/path)
+func (sd *ServiceDetails) GetEndpointURL() string {
+	return sd.endpointURL
+}
+
+// public endpoint
+// 1) Default (service override nil, no definition for public endpt in externalEndpoints)
+// - create route + pass in service selector
+// 2) externalEndpoints != nil for public endpt
+// - pass in override for metallb LoadBalancer service
+// 3) service override provided == LoadBalancer service type
+// - do nothing let the service operator handle it
+// 4) service override provided == ClusterIP service type
+// - create route , merge override with service selector
+//
+// internal endpoint
+// 1) Default (service override nil, no definition for internal endpt in externalEndpoints)
+// - service operator will create default service
+// 2) externalEndpoints != nil for internal endpt, no service override provided
+// - pass in override for metallb LoadBalancer service
+// 3) externalEndpoints != nil for internal endpt, service override provided
+// - merge override with metallb service and pass in override
+// 4) externalEndpoints == nil for internal endpt, service override provided
+// - do nothing let the service operator handle it
+
+// CreateEndpointServiceOverride -
+func (sd *ServiceDetails) CreateEndpointServiceOverride() (*service.Service, error) {
+
+	// get service overrides for the endpoint
+	svcOverride := service.GetOverrideSpecForEndpoint(sd.ServiceOverrideSpec, sd.Endpoint)
+
+	// get MetalLB config data for the endpoint
+	metalLBData := GetExternalEnpointDetailsForEndpoint(sd.ExternalEndpoints, sd.Endpoint)
+
+	// if there is no override and no metallb config for the internal endpoint return
+	if svcOverride == nil && metalLBData == nil && sd.Endpoint == service.EndpointInternal {
+		return nil, nil
+	}
+
+	// set the endpoint name to service name - endpoint type
+	sd.endpointName = sd.ServiceName + "-" + string(sd.Endpoint)
```

Comment:
This seems like duplicated information between here and https://github.com/openstack-k8s-operators/keystone-operator/pull/289/files#diff-89511165055af64058394ac61e20f832806f3f325b9cb3771f723d7e6cb94c8cR467

I guess we need to ensure that the two places generates the same endpointName so they should use a common code to generate that name.

Reply:
I initially moved keystone.ServiceName to be defined in the api module. we can have it there and use it in both places?

Reply:
since we do not allow to customize the service name we know the base and scheme of the service name

Reply:
And that knowledge of the implementation of the service operator creates the coupling. If we decide to change the naming scheme then today that only affects the service operator but after this PRs it will affect the openstack-operator as well.

Let's try the idea of moving the logic to the api module. While moving the service name generation logic to the api module is not optimal, as it moves more logic to the api module that we try to keep simple, it at least removes the logic duplication and at the same time makes the coupling explicit. We have to be careful though, we need to restrict the naming scheme to something that is "static", i.e every user of that api module if passes the same api struct and asking for the related service name(s) should get the same result. (e.g. the service name generation cannot contain anything that is specific for the ENV of the process or use any randomization like generateName from k8s).

//later

Lets use the most complex example I found which is nova-metadata. Today the Service name is generated with the following logic:
```golang
	serviceName := novametadata.ServiceName
	if instance.Spec.CellName != "" {
		serviceName = novametadata.ServiceName + "-" + instance.Spec.CellName
	}
```
https://github.com/openstack-k8s-operators/nova-operator/blob/3e970decb4afd31681f9e5a15526722be3fb79ea/controllers/novametadata_controller.go#L442-L445

Now this logic is generating a "static" name as the name only depends on constants (novametadata.ServiceName) and the values of the NovaMetadata.Spec (instance.Spec.CellName). To share this logic between the openstack-operator and the nova-operator this needs to be pulled out to a function in the nova-operator/api module. 

There is a complication though. The current logic in the reconciler operates on the NovaMetadataSpec but openstack-operator can only see NovaMetadataTemplate as part of the NovaSpec which are two separate structs. And the latter does not have CellName fields as that is coming from the NovaSpec.CellTemplates keys.

So the nova-operator/api module could have the function:
```golang
func (s NovaMetadataSpec) GetServiceName() string {
	serviceName := novametadata.ServiceName
	if s.CellName != "" {
		serviceName = novametadata.ServiceName + "-" + s.CellName
	}
        return serviceName
}
```
But for openstack-operator it is useless and openstack-operator does not have NovaMetadataSpec. So it is not enough to couple the name generation scheme between the two operators we also need to couple on the logic how nova-operator creates NovaMetadataSpec which is a multi step process. The common starting point is NovaSpec, both openstack-operator and nova-operator see that. But nova-operator processes that: NovaSpec.CellTemplate[x] is a NovaCellTemplate and nova-operator generates a NovaCellSpec out of it, then later a NovaMetadataSpec is generated from the NovaCellSpec. So far we considered all the Spec handling logic internal to nova-operator. The only really public thing was the NovaSpec and all the *Template structs.

Reply:
we have ` instance.Spec.Nova.Template.CellTemplates` in the openstack-operator. Is there something wrong using?
```
	for cellName, template := range instance.Spec.Nova.Template.CellTemplates {
```